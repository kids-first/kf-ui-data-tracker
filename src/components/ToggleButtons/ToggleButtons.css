.ToggleButtons {
  .Button {
    @inherit: .inline-flex, .items-center;
    border-radius: 5px;

    .Icon {
      @inherit: .pt-4;
      width: 14px;
      height: 14px;
    }
    &:first-child {
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    }
    &:last-child {
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
    }
  }
}
