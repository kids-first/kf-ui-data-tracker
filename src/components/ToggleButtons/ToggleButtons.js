import React, {useState} from 'react';
import propTypes from 'prop-types';
import {Button, Icon} from 'semantic-ui-react';

/**
 * Toggle Buttons taking two or more buttons and show as a group for toggle
 * Takes buttons prop as an array of {text, icon} objects
 * Dispaly text on buttons as optional
 * Customize diffrent buttons in diffrent sized
 */
const ToggleButtons = ({buttons, onToggle, size, hideText}) => {
  const [active, setActive] = useState(0);

  return (
    <Button.Group size={size}>
      {buttons.map(({text, icon}, idx) => (
        <Button
          key={text}
          icon={hideText}
          primary={active === idx}
          onClick={() => {
            setActive(idx);
            onToggle({active, text, icon});
          }}
          tabIndex="0"
        >
          <Icon name={icon} />
          {!hideText && text}
        </Button>
      ))}
    </Button.Group>
  );
};

ToggleButtons.propTypes = {
  buttons: propTypes.arrayOf(
    propTypes.shape({
      /** Button text (optional) */
      text: propTypes.string,
      /** kf-uikit icon name (optional) */
      icon: propTypes.string,
    }),
  ),
  /** onClick callback that rececives the button object as args */
  onToggle: propTypes.func,
  /** Button group size. */
  size: propTypes.oneOf([
    'mini',
    'tiny',
    'small',
    'medium',
    'large',
    'big',
    'huge',
    'massive',
  ]),
  /** Show toggle button as icons obly */
  hideText: propTypes.bool,
};

export default ToggleButtons;
