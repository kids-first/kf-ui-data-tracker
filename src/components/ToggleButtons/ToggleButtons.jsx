import React, {useState} from 'react';
import propTypes from 'prop-types';
import classnames from 'classnames';
import {Button} from 'kf-uikit';

const ToggleButtons = ({buttons, className, onToggle}) => {
  const ToggleButtonsClass = classnames('ToggleButtons', className);
  const [active, setActive] = useState(0);

  return (
    <div className={ToggleButtonsClass}>
      {buttons.map(({text, icon}, idx) => (
        <Button
          key={text + icon}
          icon={icon}
          color={active === idx ? 'primary' : null}
          onClick={() => {
            setActive(idx);
            onToggle({active, text, icon});
          }}
          tabIndex="0"
        >
          {text}
        </Button>
      ))}
    </div>
  );
};

ToggleButtons.propTypes = {
  buttons: propTypes.arrayOf(
    propTypes.shape({
      /** Button text (optional) */
      text: propTypes.string,
      /** kf-uikit icon name (optional) */
      icon: propTypes.string,
    }),
  ),
  /** additional classes to add the wrapping div */
  className: propTypes.string,
  /** onClick callback that rececives the button object as args */
  onToggle: propTypes.func,
};

export default ToggleButtons;
