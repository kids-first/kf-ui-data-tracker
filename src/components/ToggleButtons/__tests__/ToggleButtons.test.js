import React from 'react';
import {
  fireEvent,
  render,
  cleanup,
  waitForElement,
} from 'react-testing-library';
import ToggleButtons from '../ToggleButtons';

afterEach(cleanup);

it('renders correctly', () => {
  const mockButtons = [
    {text: 'Grid', icon: 'resources'},
    {text: 'list', icon: 'study'},
  ];
  const tree = render(
    <ToggleButtons onToggle={jest.fn()} buttons={mockButtons} />,
  );
  const activeButton = tree.getByText(mockButtons[0].text);
  // console.log(activeButton.className);
  expect(tree.container).toMatchSnapshot();
  // only renders one active button
  expect(activeButton.className).toBe('Button Button--default Button--primary');
});

it('changes active button on click', async () => {
  const mockButtons = [
    {text: 'Grid', icon: 'resources'},
    {text: 'list', icon: 'study'},
  ];
  const tree = render(
    <ToggleButtons onToggle={jest.fn()} buttons={mockButtons} />,
  );
  fireEvent.click(tree.getByText('list'));
  const activeButton = await waitForElement(() =>
    tree.getByText(mockButtons[1].text),
  );
  expect(tree.container).toMatchSnapshot();
  // renders the list button with the active state
  expect(activeButton.className).toBe('Button Button--default Button--primary');
});

it('calls onToggle prop on click', async () => {
  const mockOnToggle = jest.fn();
  const mockButtons = [
    {text: 'Grid', icon: 'resources'},
    {text: 'list', icon: 'study'},
  ];
  const tree = render(
    <ToggleButtons onToggle={mockOnToggle} buttons={mockButtons} />,
  );
  fireEvent.click(tree.getByText('list'));
  const activeButton = await waitForElement(() =>
    tree.getByText(mockButtons[1].text),
  );
  expect(tree.container).toMatchSnapshot();
  // only gets called once
  expect(mockOnToggle.mock.calls.length).toBe(1);
  // returns an object with key value pair of {text: 'list'}
  expect(mockOnToggle.mock.calls[0][0].text).toBe('list');
  // renders the list button with the active state
  expect(activeButton.className).toBe('Button Button--default Button--primary');
});
