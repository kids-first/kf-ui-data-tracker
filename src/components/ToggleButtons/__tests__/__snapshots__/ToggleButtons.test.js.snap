// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls onToggle prop on click 1`] = `
<div>
  <div
    class="ui buttons"
  >
    <button
      class="ui button"
      data-testid="grid"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="grid layout icon"
      />
      Grid
    </button>
    <button
      class="ui primary button"
      data-testid="list"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="list icon"
      />
      List
    </button>
  </div>
</div>
`;

exports[`changes active button on click 1`] = `
<div>
  <div
    class="ui buttons"
  >
    <button
      class="ui primary button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="grid layout icon"
      />
      Grid
    </button>
    <button
      class="ui primary button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="list icon"
      />
      list
    </button>
  </div>
</div>
`;

exports[`renders toggle button correctly 1`] = `
<div>
  <div
    class="ui buttons"
  >
    <button
      class="ui primary button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="grid layout icon"
      />
      Grid
    </button>
    <button
      class="ui primary button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="list icon"
      />
      list
    </button>
  </div>
</div>
`;

exports[`renders toggle button hiding Text 1`] = `
<div>
  <div
    class="ui buttons"
  >
    <button
      class="ui icon primary button"
      data-testid="grid"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="grid layout icon"
      />
    </button>
    <button
      class="ui icon button"
      data-testid="list"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="list icon"
      />
    </button>
  </div>
</div>
`;

exports[`renders toggle button in diffrent sizes 1`] = `
<div>
  <div
    class="ui medium buttons"
  >
    <button
      class="ui primary button"
      data-testid="grid"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="grid layout icon"
      />
      Grid
    </button>
    <button
      class="ui button"
      data-testid="list"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="list icon"
      />
      List
    </button>
  </div>
  <div
    class="ui large buttons"
  >
    <button
      class="ui primary button"
      data-testid="grid"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="grid layout icon"
      />
      Grid
    </button>
    <button
      class="ui button"
      data-testid="list"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="list icon"
      />
      List
    </button>
  </div>
</div>
`;
